### PR TITLE
simplifies and updates logic specifically around RX

### PR DESCRIPTION
**Pull request by Holden3148**

### CODE REVIEW

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 1]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
...
-                .subscribe(pokemon -> {
-                    view.hideProgressBar();
-                    view.showPokemonSprite(pokemon.getSpriteUrl());
-                    view.showPokemonName(pokemon.getPokemonName());
-                    view.showPokemonHeight("Height - " + pokemon.getPokemonHeight());
-                    view.showPokemonWeight("Weight - " + pokemon.getPokemonWeight());
-                }, throwable -> {
-                    performNetworkCall(pokemonName);
-                }));
+    private Observable<PokemonApi.Pokemon> fetchFromNetwork(String name) {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:59]: Wrapper around the pokemonService that saves the pokemon to disk as a side effect

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 2]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
         view.showProgressBar();
         view.clearPokemonInput();
 
-        checkDatabase(pokemonName);
+        findPokemon(pokemonName);
+    }
+
+    private Observable<PokemonApi.Pokemon> findInDatabase(String name) {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:55]: Wrapper around pokemonDatabaseService that turns the single into an observable to make it more cohesive

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 3]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
...
-                }, throwable -> {
-                    performNetworkCall(pokemonName);
-                }));
+    private Observable<PokemonApi.Pokemon> fetchFromNetwork(String name) {
+        return pokemonService.getPokemon(name)
+                // save pokemon to database for next call
+                .doOnNext(pokemonDatabaseService::insertPokemon);
+    }
 
+    private Observable<PokemonApi.Pokemon> findInDatabaseElseFetchFromNetwork(String name) {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:65]: Helper function that tries to read from db first, if that fails or is empty switch over to the network that will cache the value to the database for the next call

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 4]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
...
+        return findInDatabase(name)
+                // if error, return empty observable
+                .onErrorResumeNext(Observable.empty())
+                // switch to network if empty
+                .switchIfEmpty(fetchFromNetwork(name));
+    }
 
+    private Consumer<PokemonApi.Pokemon> updateUiForPokemon() {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:73]: Single method for handling UI updates

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 5]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
...
-                    view.showPokemonSprite(pokemon.getSpriteUrl());
-                    view.showPokemonName(pokemon.getPokemonName());
-                    view.showPokemonHeight("Height - " + pokemon.getPokemonHeight());
-                    view.showPokemonWeight("Weight - " + pokemon.getPokemonWeight());
-                    pokemonDatabaseService.insertPokemon(pokemon);
-                }, throwable -> {
-                    view.hideProgressBar();
-                    view.showErrorToast();
-                }));
+    private Consumer<Throwable> handlerError() {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:86]: Single method for handling error cases

File: app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java [hunk 6]
```diff
@@ -53,39 +49,48 @@ public void searchClicked(String pokemonName) {
...
+        return throwable -> {
+            view.hideProgressBar();
+            view.showErrorToast();
+        };
     }
 
+    public void findPokemon(String name) {
```
Comment (Holden3148) [inline app/src/main/java/com/elkcreek/rodneytressler/intermediatepokedex/ui/PokedexView/PokedexPresenter.java:93]: function that brings it all together for finding a pokemon

**Review by Travis40508 — APPROVED** [review]

**Pull request merged**